Improving CompactionManager using cost benefit analysis

nsivabalan opened the pull request:
- This patch improves the CompactionManager to determine the best candidate to compact using cost benefit analysis.
- Added CompactionPolicies(and factories) which assists in determining the candidate to compact.
1. CompactAll : returns entire log segment range to be compacted
2. Default        : Using simple cost benefit analysis, find the best candidate to compact. 
- Have restructured the APIs a bit so that BlobStore doesn't need to expose unnecessary apis. Hence, the BlobStore calls into CompactionPolicy.getCompactionDetails(....) by passing in required arguments.

Design: https://github.com/linkedin/ambry/issues/570

SLA: 40 mins
Reviewers: @cgtz @vgkholla 

Build succeeded in both OS. Coding style applied

Coverage:
Package	                                         Class, %	         Method, %	  Line, %
CompactAllPolicy	                        100% (1/ 1)	100% (2/ 2)	100% (9/ 9)
CompactAllPolicyFactory              100% (1/ 1)	100% (2/ 2)	100% (4/ 4)
CompactionManager	                100% (1/ 1)	100% (2/ 2)	100% (8/ 8)
CostBenefitInfo	                        100% (1/ 1)	87.5% (7/ 8)	94.1% (16/ 17)
DefaultCompactionPolicy	        100% (1/ 1)	100% (4/ 4)	100% (56/ 56)
DefaultCompactionPolicyFactory	100% (1/ 1)	100% (2/ 2)	100% (4/ 4)

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+            invalidLogSegments.add(segmentName);
+          }
+        }
+        for (String invalidLogSegment : invalidLogSegments) {
```
if you use an iterator, you could remove in a single loop. There will be no need to have the temporary `invalidLogSegments` variable.

nsivabalan replied:
I got concurrentModificationException and hence had to do it this way

vgkholla replied:
When using an iterator?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+        Map.Entry<String, Long> firstEntry = validDataPerLogSegments.getSecond().firstEntry();
+        Map.Entry<String, Long> lastEntry = validDataPerLogSegments.getSecond().lastEntry();
+        CostBenefitInfo bestCandidateToCompact = null;
+        while (!firstEntry.getKey().equals(lastEntry.getKey())) {
```
is the ending condition first == last or first > last? first == last seems like a valid case for calculating cost benefit right? That is case where a single log segment may have no valid entries at all and is eligible to be completely dropped. That might be the best cost benefit we get (> 1 and 0 cost).

vgkholla replied:
If this is a valid case, please also add a test for it.

vgkholla replied:
In fact if the cost benefit ratio is 0 (i.e no cost), you should probably compare the actual benefit. You could probably implement comparable in `CostBenefitRatio` taking all of these into account.

nsivabalan replied:
good catch. Have fixed it and have added tests as well.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+          Map.Entry<String, Long> endEntry = lastEntry;
+          while (!firstEntry.getKey().equals(endEntry.getKey())) {
+            CostBenefitInfo costBenefitInfo =
+                getCostBenefitInfo(firstEntry.getKey(), endEntry.getKey(), validDataPerLogSegments.getSecond(),
```
minor: could you store `validDataPerLogSegments.getSecond()` in a variable? Just for readability.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+                    blobStore.getSegmentCapacity(), blobStore.getLogSegmentHeaderSize(),
+                    blobStoreStats.getErrorMargin());
+            // @TODO: introduce storeConfig for minimum benefit to consider
+            if (costBenefitInfo.benefit > 1 && (bestCandidateToCompact == null
```
> 0 or > 1? You want at least two segments to be reclaimed?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+            if (costBenefitInfo.benefit > 1 && (bestCandidateToCompact == null
+                || costBenefitInfo.costBenefitRatio < bestCandidateToCompact.costBenefitRatio)) {
+              bestCandidateToCompact = costBenefitInfo;
+              logger.trace("Updating best candidate to compact to " + bestCandidateToCompact);
```
please use log formatting

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+      long errorMargin) {
+    // cost = total valid data size (or data that needs to be copied over)
+    // benefit = no of segments reclaimed = (total source segments - total target segments)
+    //           total target segments = Ceil (total cost / (segment capacity - segment header - error margin))
```
minor: extra spaces ? (not sure if diff or there are actually spaces after the //)

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+    String curSegmentName = firstLogSegmentName;
+    while (!curSegmentName.equals(lastLogSegmentName)) {
+      totalCost += validDataSizePerLogSegment.get(curSegmentName);
+      curSegmentName = validDataSizePerLogSegment.higherEntry(curSegmentName).getKey();
```
can do `higherKey()`?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+    totalCost += validDataSizePerLogSegment.get(curSegmentName);
+    totalSegmentsToBeCompacted++;
+    long maxCapacityPerSegment = segmentCapacity - segmentHeaderSize - errorMargin;
+    int benefit = totalSegmentsToBeCompacted - (int) Math.ceil(totalCost / (maxCapacityPerSegment * 1.0));
```
is the `* 1.0` required?

nsivabalan replied:
yes, we want to find the actual double value of (cost / capacity) (which is long/ long) and then take ceil over it.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+      this.lastLogSegmentName = lastLogSegmentName;
+      this.cost = cost;
+      this.benefit = benefit;
+      costBenefitRatio = benefit == 0 ? cost : cost / benefit;
```
shouldn't this be `costBenefitRatio = benefit == 0 ? Double.MAX_VALUE : cost / benefit;`? Otherwise there would not be a difference b/w benefit = 0 and benefit = 1.

Also, you need to cast `cost` to `double` - otherwise the ratio you get is an integer (i.e 20/7 will result in 2.0).

nsivabalan replied:
got it

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -47,9 +54,109 @@ CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreExceptio
+        // fetch valid data size from BlobStoreStats
+        BlobStoreStats blobStoreStats = blobStore.getBlobStoreStats();
+        Pair<Long, NavigableMap<String, Long>> validDataPerLogSegments = blobStoreStats.getValidDataSizeByLogSegment(
+            new TimeRange(time.milliseconds() - messageRetentionTimeInMs, 0));
```
yes, for now, all stats requests are determined on demand by the BlobStoreStats. So, once we have bucketing, we can revisit this.

nsivabalan commented on ambry-store/src/main/java/com.github.ambry.store/BlobStore.java:
```diff
@@ -371,14 +373,38 @@ long getCapacityInBytes() {
+   * Fetches segment header size of a single log segment
+   * @return the segment header size of a single log segment
+   */
+  long getLogSegmentHeaderSize() {
```
Makes sense. I am looking at something like this. Lets discuss offline and take a call.

```
interface CostBenefitCalculator {

  CompactionDetails getCompactionDetails(long totalCapacity, long usedCapacity, long segmentCapacity,
      long segmentHeaderSize, List<String> logSegmentsNotInJournal, BlobStoreStats blobStoreStats)
      throws StoreException;

  // goes over all candidates who doesn't over lap with jounral and finds the best candidate
  CostBenefitInfo getBestCandidateToCompact(NavigableMap<String, Long> validDataPerLogSegments, long segmentCapacity,
      long segmentHeaderSize, long errorMargin);

  // finds the cost benefit ratio for a given candidate
  CostBenefitInfo getCostBenefitInfo(String firstLogSegmentName, String lastLogSegmentName,
      NavigableMap<String, Long> validDataSizePerLogSegment, long segmentCapacity, long segmentHeaderSize,
      long errorMargin);
}

class SimpleCostBenefitCalculator implements CostBenefitCalculator{

 // all method definitions
 .
 .
 .
 }

In CompactionManager: 
SimpleCostBenefitCalculator calculator = new SimpleCostBenefitCalculator()
CompactionDetails details = store.computeCostBenefitAndGetCompactionDetails(calculator);

In BlobStore:
CompactionDetails computeCostBenefitAndGetCompactionDetails(CostBenefitCalculator calculator){
    return calculator.getCompactionDetails(totalCapacity(),usedCapacity, segmentCapacity,
      segmentHeaderSize, logSegmentsNotInJournal, blobStoreStats);
}
```

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CompactAllPolicy.java:
```diff
@@ -0,0 +1,48 @@
+ * CompactAllPolicy returns the entire list of all log segments non overlapping with {@link Journal} as a candidate
+ * to be compacted for {@link #getCompactionDetails(long, long, long, long, List, BlobStoreStats)}
+ */
+class CompactAllPolicy implements CompactionPolicy {
```
This is the same as the v0 compaction policy, right?

nsivabalan replied:
yes, you are right

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -41,15 +38,13 @@
-      if (potentialLogSegments != null) {
-        details = new CompactionDetails(time.milliseconds() - messageRetentionTimeInMs, potentialLogSegments);
-      }
+    CompactionPolicyFactory compactionPolicyFactory;
```
The instantiation of `CompactionPolicyFactory` could happen in the constructor, right? Since the factory type won't change while running

vgkholla replied:
+1

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -41,15 +38,13 @@
+      throw new IllegalStateException("Error creating compaction policy for store " + blobStore.toString());
     }
-    return details;
+    return blobStore.getCompactionDetails(compactionPolicyFactory.getCompactionPolicy());
```
Also, do you envision a scenario where a new `CompactionPolicy` instance has to be constructed each time?

vgkholla replied:
+1 - seems like a single instance should be enough, the factory should be able to decide whether a new instance is required or a single instance is good enough.

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CompactionPolicyFactory.java:
```diff
@@ -0,0 +1,25 @@
+interface CompactionPolicyFactory {
+
+  /**
+   * @return the {@link CompactionPolicy} to be use
```
"to be use" to "to use"

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CostBenefitInfo.java:
```diff
@@ -0,0 +1,85 @@
+  private final String firstLogSegmentName;
+  private final String lastLogSegmentName;
+
+  CostBenefitInfo(String firstLogSegmentName, String lastLogSegmentName, long cost, int benefit) {
```
why is cost a `long` and benefit an `int`?

nsivabalan replied:
Cost is total amount of data that needs to be copied in bytes(and hence long). Benefit is no of log segments reclaimed and hence an integer.

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CostBenefitInfo.java:
```diff
@@ -0,0 +1,85 @@
+    this.lastLogSegmentName = lastLogSegmentName;
+    this.cost = cost;
+    this.benefit = benefit;
+    costBenefitRatio = benefit == 0 ? Double.MAX_VALUE : cost * 1.0 / benefit;
```
minor: I think `(float) cost / benefit` or `cost * (1.0 / benefit)` looks a bit cleaner

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java:
```diff
@@ -39,6 +39,8 @@
 class BlobStoreStats implements StoreStats {
   static final String IO_SCHEDULER_JOB_TYPE = "BlobStoreStats";
   static final String IO_SCHEDULER_JOB_ID = "indexSegment_read";
+  // Max blob size to factor in while calculating the cost benefit
```
comment is out of place - no cost-benefit here.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/BlobStoreStats.java:
```diff
@@ -50,6 +52,14 @@
   }
 
+  /**
+   * Returns the max blob size to factor in while calculating the cost benefit
```
same here, cost benefit is the terminology for a compaction policy.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactAllPolicy.java:
```diff
@@ -0,0 +1,48 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.store;
```
Maybe you can just make this in an internal class in the factory.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CostBenefitInfo.java:
```diff
@@ -0,0 +1,85 @@
+  private final long cost;
+  private final int benefit;
+  private final Double costBenefitRatio;
+  private final String firstLogSegmentName;
```
why not store the list of log segments? You would have had to cycle through them to get the cost benefit.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CostBenefitInfo.java:
```diff
@@ -0,0 +1,85 @@
+  @Override
+  public int compareTo(CostBenefitInfo that) {
+    if (costBenefitRatio.compareTo(that.costBenefitRatio) != 0) {
+      return costBenefitRatio.compareTo(that.costBenefitRatio);
```
seems a bit wasteful to calculate twice - would prefer calculating once and returning the value.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CostBenefitInfo.java:
```diff
@@ -0,0 +1,85 @@
+  /**
+   * @return the cost benefit ratio associated with this {@link CostBenefitInfo}
+   */
+  Double getCostBenefitRatio() {
```
are these three getters required (cost, benefit and ratio)?

nsivabalan replied:
benefit and ratio is required for DefaultCompactionPolicy. and cost is used only in tests. But to make it uniform I have exposed cost as well.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java:
```diff
@@ -0,0 +1,144 @@
+    CompactionDetails details = null;
+    if (usedCapacity >= (storeConfig.storeMinUsedCapacityToTriggerCompactionInPercentage / 100.0) * totalCapacity) {
+      if (logSegmentsNotInJournal != null) {
+        Pair<Long, NavigableMap<String, Long>> validDataPerLogSegments = blobStoreStats.getValidDataSizeByLogSegment(
```
PerLogSegment

vgkholla replied:
perhaps best to name it like the function i.e `validDataSizeByLogSegment`.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java:
```diff
@@ -0,0 +1,144 @@
+    if (usedCapacity >= (storeConfig.storeMinUsedCapacityToTriggerCompactionInPercentage / 100.0) * totalCapacity) {
+      if (logSegmentsNotInJournal != null) {
+        Pair<Long, NavigableMap<String, Long>> validDataPerLogSegments = blobStoreStats.getValidDataSizeByLogSegment(
+            new TimeRange(time.milliseconds() - messageRetentionTimeInMs, 0));
```
would recommend providing an hour of error margin here. The error margin that this impl is ok with should not depend on whether bucketing is implemented or not.

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java:
```diff
@@ -0,0 +1,144 @@
+
+        Iterator<Map.Entry<String, Long>> validEntryIterator =
+            validDataPerLogSegments.getSecond().entrySet().iterator();
+        while (validEntryIterator.hasNext()) {
```
is this loop required at all? What if we took a submap starting from the beginning to the last element in `logSegmentsNotInJournal`?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java:
```diff
@@ -0,0 +1,144 @@
+    Map.Entry<String, Long> firstEntry = validDataPerLogSegments.firstEntry();
+    Map.Entry<String, Long> lastEntry = validDataPerLogSegments.lastEntry();
+    CostBenefitInfo bestCandidateToCompact = null;
+    while (firstEntry != null && firstEntry.getKey().compareTo(lastEntry.getKey()) <= 0) {
```
not sure why the second condition is required - `lastEntry` never changes right?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/DefaultCompactionPolicy.java:
```diff
@@ -0,0 +1,144 @@
+        CostBenefitInfo costBenefitInfo =
+            getCostBenefitInfo(firstEntry.getKey(), endEntry.getKey(), validDataPerLogSegments, segmentCapacity,
+                segmentHeaderSize, maxBlobSize);
+        // @TODO: introduce storeConfig for minimum benefit to consider
```
why not introduce it right away?

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactAllPolicyTest.java:
```diff
@@ -0,0 +1,69 @@
+   */
+  public CompactAllPolicyTest() throws InterruptedException {
+    Pair<MockBlobStore, StoreConfig> initState = CompactionPolicyTest.initializeBlobStore(properties, time, -1, -1, -1);
+    this.config = initState.getSecond();
```
minor: `this` qualifiers not required.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactionManagerTest.java:
```diff
@@ -17,217 +17,136 @@
-   * Instantiates {@link CompactionManagerTest} with the required cast
-   * @throws InterruptedException
+   * Running for both CompactAll and Default Compaction Policy
+   * @return an array with both {@code CompactionPolicyType.DEFAULT} and {@code CompactionPolicyType.COMPACT_ALL}.
```
why do we need this to run with two policies? This looks like an integration test to me. If anything, we should mock the policy (or use compact all) for the tests here.

nsivabalan replied:
Sure. Will mock the policy.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactionPolicyFactoryTest.java:
```diff
@@ -0,0 +1,77 @@
+   */
+  @Test
+  public void testCompactionPolicyFactoryFailureCases()
+      throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException, InstantiationException,
```
`throws Exception`.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactionPolicyFactoryTest.java:
```diff
@@ -0,0 +1,77 @@
+    StoreConfig config = new StoreConfig(new VerifiableProperties(properties));
+    Time time = new MockTime();
+    try {
+      Utils.getObj(config.storeCompactionPolicyFactory, config, time);
```
this test is testing `Utils.getObj()` - it is not required I think.

nsivabalan replied:
yes. Will remove it.

nsivabalan commented:
@vgkholla @cgtz : Addressed all your comments. (Yet to resolve merge conflicts for CompactionManager and StoreConfig. But you guys are good to review though)

vgkholla commented on ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java:
```diff
@@ -118,6 +118,20 @@
+   */
+  @Config("store.min.log.segment.count.to.reclaim.to.trigger.compaction")
+  @Default("1")
+  public final int storeMinLogSegmentCountToReclaimToTriggerCompaction;
```
How do we make it clear that this need not be followed by all compaction policies? Perhaps you can mention that in the documentation that respecting this config is up to the implementation of the individual policy impls

vgkholla commented on ambry-api/src/main/java/com.github.ambry/config/StoreConfig.java:
```diff
@@ -141,6 +155,11 @@ public StoreConfig(VerifiableProperties verifiableProperties) {
+    storeCompactionPolicyFactory = verifiableProperties.getString("store.compaction.policy.factory",
+        "com.github.ambry.store.DefaultCompactionPolicyFactory");
+    storeMinLogSegmentCountToReclaimToTriggerCompaction =
+        verifiableProperties.getInt("store.min.log.segment.count.to.reclaim.to.trigger.compaction", 1);
```
use `getIntInRange()`

nsivabalan replied:
This would depend on the max no of log segments right. Else how do we set the max value ?

vgkholla replied:
`Integer.MAX_VALUE` - the `getIntInRange()` is to guard against negative values

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactAllPolicyFactory.java:
```diff
@@ -0,0 +1,66 @@
+
+
+/**
+ *  Factory class interface to get {@link CompactAllPolicy}
```
class interface?

vgkholla commented on ambry-store/src/main/java/com.github.ambry.store/CompactionManager.java:
```diff
@@ -98,16 +103,7 @@ void awaitTermination() {
    * @throws StoreException when {@link BlobStore} is not started
    */
   CompactionDetails getCompactionDetails(BlobStore blobStore) throws StoreException {
```
can we make this a `private` function now? It will be tested along with the `CompactionExecutor` anyway. The actual policy tests are in the respective tests.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactionManagerTest.java:
```diff
@@ -75,111 +70,43 @@ public void testEnableDisable() {
+    blobStore.exceptionToThrowOnGetCompactionDetails =
+        new StoreException("Random StoreException to test", StoreErrorCodes.Unknown_Error);
+    try {
+      compactionManager.getCompactionDetails(blobStore);
```
need a `fail()`

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactionPolicyTest.java:
```diff
@@ -0,0 +1,365 @@
+   * Refers to different types of {@link CompactionPolicy}
+   */
+  enum CompactionPolicyType {
+    DEFAULT,
```
why not use the factories instead? The parameters can be the class name of the factories (just like in config).

nsivabalan replied:
Initializing is not an issue. But in every test, we need to know what type the policy is to determine the expected output. So, I have chosen this way.

vgkholla replied:
could use `instanceof`?

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactionPolicyTest.java:
```diff
@@ -0,0 +1,365 @@
+          initializeBlobStore(properties, time, -1, messageRetentionDays, DEFAULT_MAX_BLOB_SIZE);
+      if (type.equals(CompactionPolicyType.DEFAULT)) {
+        bestCandidates = setUpStateForDefaultCompactionPolicy(blobStore, mockBlobStoreStats);
+        compactionPolicy = new DefaultCompactionPolicy(initState.getSecond(), time);
```
isn't this already available as a class variable?

nsivabalan replied:
we initialize a new StoreConfig within the for loop. and hence need to construct a new compactionPolicy

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/DefaultCompactionPolicyTest.java:
```diff
@@ -0,0 +1,226 @@
+  public DefaultCompactionPolicyTest() throws InterruptedException {
+    Pair<MockBlobStore, StoreConfig> initState =
+        CompactionPolicyTest.initializeBlobStore(properties, time, -1, -1, DEFAULT_MAX_BLOB_SIZE);
+    this.config = initState.getSecond();
```
minor: `this` qualifiers not required.

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/DefaultCompactionPolicyTest.java:
```diff
@@ -0,0 +1,226 @@
+              CompactionPolicyTest.generateValidDataSize(blobStore.logSegmentsNotInJournal, bestCandidates, bestCost,
+                  maxLogSegmentCapacity);
+          mockBlobStoreStats.validDataSizeByLogSegments = validDataSize;
+          if ((j - i) > 0) { // only if benefit is > 0. TODO: fix once the storeConfig is added
```
can the TODO be done now that the config is introduced?

cgtz commented on ambry-store/src/test/java/com.github.ambry.store/DefaultCompactionPolicyTest.java:
```diff
@@ -0,0 +1,226 @@
+   */
+  @Test
+  public void testGetCompactionDetailsTest() throws StoreException, InterruptedException {
+    // null logSegmentsNotInJournal
```
Could this comment explain why this is set to null here (and then set again a few lines below)?

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CostBenefitInfo.java:
```diff
@@ -0,0 +1,80 @@
+    this.segmentsToCompact = segmentsToCompact;
+    this.cost = cost;
+    this.benefit = benefit;
+    costBenefitRatio = benefit == 0 ? Double.MAX_VALUE : cost * (1.0 / benefit);
```
`Double.MAX_VALUE` is the largest finite double, not infinity. Is that what you intended here?

nsivabalan replied:
yes.

cgtz commented on ambry-store/src/main/java/com.github.ambry.store/CompactAllPolicyFactory.java:
```diff
@@ -0,0 +1,66 @@
+ * CompactAllPolicy returns the entire list of all log segments non overlapping with {@link Journal} as a candidate
+ * to be compacted for {@link #getCompactionDetails(long, long, long, long, List, BlobStoreStats)}
+ */
+class CompactAllPolicy implements CompactionPolicy {
```
I guess this could also be a private static class inside of CompactAllPolicyFactory, as well.

vgkholla replied:
+1 could it be `static`?

vgkholla replied:
or you could avoid having the repeated class variables (`storeConfig` and `time`) and not make it `static`

nsivabalan commented:
@vgkholla @cgtz : addressed all comments

vgkholla commented on ambry-store/src/test/java/com.github.ambry.store/CompactionPolicyTest.java:
```diff
@@ -46,50 +47,41 @@
   private static long DEFAULT_USED_CAPACITY_IN_BYTES = CAPACITY_IN_BYTES * 6 / 10;
   private static long DEFAULT_MAX_BLOB_SIZE = CAPACITY_IN_BYTES / 100;
   private static long SEGMENT_HEADER_SIZE = CAPACITY_IN_BYTES / 50;
+  private static String COMPACT_ALL_POLICY_FACTORY = "com.github.ambry.store.CompactAllPolicyFactory";
```
`final` on all of these

cgtz replied:
Does `CompactAllPolicyFactory.class.getName()` work?

vgkholla replied:
+1
